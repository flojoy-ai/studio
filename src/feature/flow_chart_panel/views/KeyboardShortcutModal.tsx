import { createStyles, Modal } from "@mantine/core";
import { memo, useState } from "react";
import { Button } from "@/components/ui/button";
import { Dialog, DialogContent, DialogHeader } from "@/components/ui/dialog";
import { platform } from "os";
import { ScrollArea } from "@src/components/ui/scroll-area";

interface KeyboardShortcutProps {
  handleKeyboardShortcutModalOpen: (open: boolean) => void;
  isKeyboardShortcutModalOpen: boolean;
}

//all existing styles in this file should be made using createStyles

const useStyles = createStyles((theme) => ({
  content: {
    borderRadius: "8px",
    height: "85vh",
    width: "max(400px,936px)",
    position: "relative",
    inset: 0,
    padding: 0,
  },
  overlay: {
    zIndex: 100,
    top: 0,
    left: 0,
    height: "100%",
    width: "100%",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
  },
  closeButton: {
    position: "absolute",
    backgroundColor: "transparent",
    border: 0,
    cursor: "pointer",
    top: 15,
    right: 10,
    padding: 0,
    color: theme.colors.accent1[0],
  },
  container: {
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    gap: 43,
    height: "100%",
    width: "100%",
    padding: 24,
    backgroundColor: theme.colors.modal[0],
  },
  column: {
    width: "100%",
  },
  title: {
    fontSize: 24,
    fontWeight: "bold",
    fontFamily: "Inter",
    marginBottom: 10,
  },
  platformName: {
    color: "#3d7ff2",
  },
  list: {
    display: "flex",
    flexDirection: "column",
    alignItems: "flex-start",
    padding: theme.spacing.md, // 24px
    gap: theme.spacing.xs, // 8px
    boxShadow: "0px 0px 20px rgba(0, 0, 0, 0.15)",
    borderRadius: theme.radius.md, // 8px
    backgroundColor: theme.colors.modal[0],
    color: theme.colors.text[0],
    border: `1px solid ${theme.colors.modal[0]}`,
    width: "100%",
  },
  listItem: {
    display: "flex",
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "flex-start",
    padding: theme.spacing.xs, // 8px
    width: "100%",
    boxShadow: "0px 0px 20px rgba(0, 0, 0, 0.1)",
    borderRadius: theme.radius.xs, // 2px
  },
  commandKey: {
    color: "#3d7ff2",
  },
}));

const KeyboardShortcutModal = ({
  handleKeyboardShortcutModalOpen,
  isKeyboardShortcutModalOpen,
}: KeyboardShortcutProps) => {
  const { classes } = useStyles();
  // const [isKeyboardShortcutModalOpen, setIsKeyboardShortcutOpen] =
  //   useState<boolean>(false);

  // return (
  //   <Modal
  //     data-testid="keyboard_shortcut_modal"
  //     opened={isOpen}
  //     onClose={onClose}
  //     size={1030}
  //   >
  //     <Button
  //       data-testid="keyboard_shortcut-closebtn"
  //       onClick={onClose}
  //       className={classes.closeButton}
  //     ></Button>

  //     <div data-testid="key_container" className={classes.container}>
  //       {platforms.map((platform) => {
  //         return (
  //           <div className={classes.column} key={platform.key}>
  //             <div className={classes.title}>
  //               For{" "}
  //               <span className={classes.platformName}>{platform.title}</span>
  //             </div>

  //             <div className={classes.list}>
  //               {keyboardShortcuts.map((shortcut) => (
  //                 <div className={classes.listItem} key={shortcut.command}>
  //                   <span>{shortcut.command}</span>
  //                   <span className={classes.commandKey}>
  //                     {shortcut.platforms[platform.key]}
  //                   </span>
  //                 </div>
  //               ))}
  //             </div>
  //           </div>
  //         );
  //       })}
  //     </div>
  //   </Modal>
  // );
  return (
    <Dialog
      open={isKeyboardShortcutModalOpen}
      onOpenChange={handleKeyboardShortcutModalOpen}
    >
      <DialogContent className="h-[560px] max-w-5xl">
        <DialogHeader className="mb-2"></DialogHeader>
        <ScrollArea>
          <div data-testid="key_container" className="flex justify-center">
            {platforms.map((platform) => {
              return (
                <div className="w-full" key={platform.key}>
                  <div className=" mb-4 text-2xl font-bold">
                    <span className=" text-gray-200">For</span>{" "}
                    <span className=" text-blue-500">{platform.title}</span>
                  </div>

                  <div className="border-1 flex w-full flex-col items-start gap-2 rounded-lg border-solid border-modal bg-modal p-6 text-gray-200 shadow-lg">
                    {keyboardShortcuts.map((shortcut) => (
                      <div
                        className="flex w-full flex-row items-start justify-between rounded-sm p-3 shadow-xl"
                        key={shortcut.command}
                      >
                        <span>{shortcut.command}</span>
                        <span className="text-blue-500">
                          {shortcut.platforms[platform.key]}
                        </span>
                      </div>
                    ))}
                  </div>
                </div>
              );
            })}
          </div>
        </ScrollArea>
      </DialogContent>
    </Dialog>
  );
};

export default memo(KeyboardShortcutModal);

const platforms = [
  { title: "Windows", key: "windows" },
  { title: "MacOs", key: "macOs" },
];

const keyboardShortcuts = [
  {
    command: "Show/hide UI",
    platforms: {
      windows: "Ctrl \\",
      macOs: "⌘ z",
    },
  },
  {
    command: "Play",
    platforms: {
      windows: "Ctrl P",
      macOs: "⌘ P",
    },
  },
  {
    command: "Open Side Bar Panel",
    platforms: {
      windows: "Ctrl T",
      macOs: "⌘ T",
    },
  },
  {
    command: "Fit View",
    platforms: {
      windows: "Ctrl 1",
      macOs: "⌘ 1",
    },
  },
  {
    command: "Zoom In",
    platforms: {
      windows: "Ctrl +",
      macOs: "⌘ +",
    },
  },
  {
    command: "Zoom Out",
    platforms: {
      windows: "Ctrl -",
      macOs: "⌘ -",
    },
  },
  {
    command: "Layout Node",
    platforms: {
      windows: "Ctrl L",
      macOs: "⌘ L",
    },
  },
  {
    command: "Select All",
    platforms: {
      windows: "Ctrl A",
      macOs: "⌘ A",
    },
  },
  {
    command: "Deselect",
    platforms: {
      windows: "Alt C",
      macOs: "Option C",
    },
  },
  {
    command: "Close Node",
    platforms: {
      windows: "Ctrl W",
      macOs: "⌘ W",
    },
  },
  {
    command: "File Name",
    platforms: {
      windows: "Ctrl R",
      macOs: "⌘ R",
    },
  },
  {
    command: "Save",
    platforms: {
      windows: "Ctrl S",
      macOs: "⌘ S",
    },
  },
  {
    command: "Delete",
    platforms: {
      windows: "Ctrl D",
      macOs: "Backspace",
    },
  },
];
