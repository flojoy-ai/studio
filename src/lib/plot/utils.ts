export function clamp(x: number, lo: number, hi: number) {
  return Math.min(Math.max(x, lo), hi);
}
