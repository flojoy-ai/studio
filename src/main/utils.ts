import * as http from "http";
import * as fs from "fs";
import { isIP } from "net";
import { execCommand } from "./executor";
import { Command } from "./command";
import { app, dialog } from "electron";
import { join } from "path";
import { killCaptain } from "./python";
import log from "electron-log";
import { ChildProcess } from "node:child_process";
import { Err, Result, Ok } from "../types/result";

export const isPortFree = (port: number) =>
  new Promise((resolve) => {
    const server = http
      .createServer()
      .listen(port, "127.0.0.1", () => {
        server.close();
        resolve(true);
      })
      .on("error", () => {
        resolve(false);
      });
  });

export const killProcess = async (port: number) => {
  await execCommand(
    new Command({
      darwin: `kill -9 $(lsof -t -i :${port})`,
      linux: `kill -9 $(lsof -t -i :${port})`,
      win32: `FOR /F "tokens=5" %i IN ('netstat -aon ^| find "${port}"') DO Taskkill /F /PID %i`,
    }),
  );
};

export const ping = async (addr: string) => {
  if (isIP(addr) === 0) {
    throw new Error(`Invalid IP address: ${addr}`);
  }

  return await execCommand(
    new Command({
      win32: `ping -n 1 ${addr}`,
      darwin: `ping -c 1 ${addr}`,
      linux: `ping -c 1 ${addr}`,
    }),
    { quiet: true },
  );
};

export const netstat = async () => {
  return await execCommand(
    new Command({
      win32: "netstat",
      darwin: "netstat",
      linux: "netstat",
    }),
    { quiet: true },
  );
};

export const ifconfig = async () => {
  return await execCommand(
    new Command({
      win32: "ipconfig /all",
      darwin: "ifconfig",
      linux: "ifconfig",
    }),
    { quiet: true },
  );
};

export const writeFileSync = (_, filePath: string, text: string): void => {
  fs.writeFileSync(filePath, text);
};

export const pickDirectory = async (): Promise<string> => {
  const handler = await dialog.showOpenDialog({
    properties: ["openDirectory"],
  });
  return handler.canceled ? "" : handler.filePaths[0];
};

export const getCustomBlocksDir = async () => {
  const filePath = join(app.getPath("home"), ".flojoy/custom_blocks_path.txt");

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  const blocksPath = fs
    .readFileSync(filePath, { encoding: "utf-8" })
    .toString();

  if (!fs.existsSync(blocksPath)) {
    return undefined;
  }

  return blocksPath;
};

export const cacheCustomBlocksDir = (_, dirPath: string) => {
  const flojoyDir = join(app.getPath("home"), ".flojoy");
  if (!fs.existsSync(flojoyDir)) {
    fs.mkdirSync(flojoyDir);
  }
  const cacheFilePath = join(flojoyDir, "custom_blocks_path.txt");
  fs.writeFileSync(cacheFilePath, dirPath);
};

export const openTestPicker = (): Promise<
  { filePath: string; fileContent: string } | undefined
> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: ["json", "py"],
            name: "Test",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const openFilePicker = (
  _,
  allowedExtensions: string[] = ["json"],
): Promise<{ filePath: string; fileContent: string } | undefined> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: allowedExtensions,
            name: "allowed extensions",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
        resolve(undefined);
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const cleanup = async () => {
  const captainProcess = global.captainProcess as ChildProcess;
  log.info(
    "Cleanup function invoked, is captain running? ",
    !(captainProcess?.killed ?? true),
  );
  if (captainProcess && captainProcess.exitCode === null) {
    const success = killCaptain();
    if (success) {
      global.captainProcess = null;
      log.info("Successfully terminated captain :)");
    } else {
      log.error("Something went wrong when terminating captain!");
    }
  }
};

export const loadFileFromFullPath = async (
  filePath: string,
): Promise<string> => {
  return fs.readFileSync(filePath, { encoding: "utf-8" }).toString();
};

export const saveFileToFullPath = async (
  filePath: string,
  content: string,
): Promise<Result<void>> => {
  try {
    fs.writeFileSync(filePath, content);
    return Ok(undefined);
  } catch (error) {
    return Err(error as Error);
  }
};
