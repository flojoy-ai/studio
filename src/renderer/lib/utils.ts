import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";
import { z } from "zod";
import * as E from "fp-ts/Either";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tryParse<Z extends z.ZodTypeAny>(z: Z) {
  return (val: unknown) => {
    const res = z.safeParse(val);
    if (res.success) {
      return E.right(res.data as z.infer<Z>);
    }
    return E.left(res.error);
  };
}
