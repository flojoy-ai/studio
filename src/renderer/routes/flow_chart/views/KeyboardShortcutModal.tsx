import { memo } from "react";
import {
  Dialog,
  DialogContent,
  DialogHeader,
  DialogTitle,
} from "@/renderer/components/ui/dialog";
import { ScrollArea } from "@/renderer/components/ui/scroll-area";

type Props = {
  isKeyboardShortcutOpen: boolean;
  setIsKeyboardShortcutOpen: (val: boolean) => void;
};

const KeyboardShortcutModal = ({
  isKeyboardShortcutOpen,
  setIsKeyboardShortcutOpen,
}: Props) => {
  return (
    <Dialog
      open={isKeyboardShortcutOpen}
      onOpenChange={setIsKeyboardShortcutOpen}
    >
      <DialogContent className="max-w-sm">
        <DialogHeader>
          <DialogTitle>Keyboard Shortcuts</DialogTitle>
        </DialogHeader>
        <ScrollArea>
          <div data-testid="key_container" className="flex justify-center">
            <div className="w-full">
              <div className="flex flex-col gap-2">
                {keyboardShortcuts.map((shortcut) => (
                  <div
                    className="flex w-full flex-row py-2"
                    key={shortcut.command}
                  >
                    <span>{shortcut.command}</span>
                    <div className="grow" />
                    <span className="text-blue-500">{shortcut.key}</span>
                  </div>
                ))}
              </div>
            </div>
          </div>
        </ScrollArea>
      </DialogContent>
    </Dialog>
  );
};

export default memo(KeyboardShortcutModal);

const keyboardShortcuts = [
  {
    command: "Play",
    key: "Ctrl/⌘ P",
  },
  {
    command: "Open Side Bar Panel",
    key: "Ctrl/⌘ K",
  },
  {
    command: "Zoom In",
    key: "Ctrl/⌘ +",
  },
  {
    command: "Zoom Out",
    key: "Ctrl/⌘ -",
  },
  {
    command: "Save",
    key: "Ctrl/⌘ S",
  },
];
