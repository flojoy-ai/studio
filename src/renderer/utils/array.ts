export function filterMap<T, U>(
  arr: Array<T>,
  mapPred: (x: T) => U | undefined | null,
) {
  return arr.reduce((acc: Array<U>, x: T) => {
    const val = mapPred(x);
    if (val !== undefined && val !== null) {
      acc.push(val);
    }

    return acc;
  }, []);
}
