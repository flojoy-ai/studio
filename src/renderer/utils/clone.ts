import { DeepMutable } from "@/renderer/types/util";

export const deepMutableClone = <T>(obj: T): DeepMutable<T> => {
  return JSON.parse(JSON.stringify(obj)) as DeepMutable<T>;
};
