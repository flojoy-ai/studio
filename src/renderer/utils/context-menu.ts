export const calculateContextMenuOffset = (
  clickX: number,
  clickY: number,
  flow: HTMLDivElement,
) => {
  // Calculate position of the context menu. We want to make sure it
  // doesn't get positioned off-screen.
  const pane = flow.getBoundingClientRect();

  const contextMenuHeight = 200;
  const paneToBlock = clickY - contextMenuHeight;

  let top: number | undefined = undefined;
  let bottom: number | undefined = undefined;

  if (paneToBlock < contextMenuHeight / 2) {
    top = paneToBlock;
  } else if (paneToBlock < contextMenuHeight) {
    if (pane.height - paneToBlock < contextMenuHeight) {
      top = contextMenuHeight - paneToBlock;
    } else {
      top = paneToBlock;
    }
  } else if (pane.height - paneToBlock < contextMenuHeight) {
    top = undefined;
    bottom = pane.height - paneToBlock;
  } else {
    top = paneToBlock;
  }

  return {
    top,
    left: clickX < pane.width - 200 ? clickX : undefined,
    right: clickX >= pane.width - 200 ? pane.width - clickX : undefined,
    bottom,
  };
};
