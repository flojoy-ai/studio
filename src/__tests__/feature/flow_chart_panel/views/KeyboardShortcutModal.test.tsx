import { fireEvent, getAllByText, getByTestId } from "@testing-library/react";
import KeyboardShortcutModal from "@src/feature/flow_chart_panel/views/KeyboardShortcutModal";
import { renderWithTheme } from "@src/__tests__/__utils__/utils";

const testId = "keyboard_shortcut_modal";
// mock mantine/core Modal component

jest.mock("@mantine/core", () => ({
  __esModule: true,
  default: jest.fn(({ children }) => <div>{children}</div>),
  Modal: jest.fn(({ children }) => (
    <div data-testid="keyboard_shortcut_modal">{children}</div>
  )),
  createStyles: jest.fn(() => jest.fn(() => ({ classes: {} }))),
  Button: jest.fn(({ children, onClick }) => (
    <button onClick={onClick} data-testid="closeButton">
      {children}
    </button>
  )),
}));

describe("KeyboardShortcutModal", () => {
  const onCloseMock = jest.fn();
  beforeEach(() => {
    onCloseMock.mockClear();
  });

  it("renders correctly when isOpen is true", () => {
    const { getByTestId, getByText } = renderWithTheme(
      <KeyboardShortcutModal isOpen={true} onClose={onCloseMock} />
    );
    const modalContent = getByTestId(testId);
    expect(modalContent).toBeInTheDocument();
    expect(getByText("Windows")).toBeInTheDocument();
    expect(getByText("MacOs")).toBeInTheDocument();
    expect(modalContent).toMatchSnapshot();
  });

  it("calls onClose when the close button is clicked", () => {
    const { getByTestId } = renderWithTheme(
      <KeyboardShortcutModal isOpen={true} onClose={onCloseMock} />
    );
    const closeButton = getByTestId("closeButton");
    fireEvent.click(closeButton);
    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  it("renders the correct keyboard shortcuts for each platform", () => {
    const { getByTestId, getByText } = renderWithTheme(
      <KeyboardShortcutModal isOpen={true} onClose={onCloseMock} />
    );
    const shortcuts = [
      {
        command: "Show/hide UI",
        platforms: {
          windows: "Ctrl \\",
          macOs: "⌘ z",
        },
      },
      {
        command: "Play",
        platforms: {
          windows: "Ctrl P",
          macOs: "⌘ P",
        },
      },
      // Add more keyboard shortcuts here
    ];
    const keyContainer = getByTestId("key_container");
    expect(keyContainer).toBeInTheDocument();
    shortcuts.forEach((shortcut) => {
      const commands = getAllByText(keyContainer, shortcut.command);
      commands.forEach((cmd) => {
        expect(cmd).toBeInTheDocument();
      });
      expect(getByText(shortcut.platforms["windows"])).toBeInTheDocument();
      expect(getByText(shortcut.platforms["macOs"])).toBeInTheDocument();
    });
  });
});
