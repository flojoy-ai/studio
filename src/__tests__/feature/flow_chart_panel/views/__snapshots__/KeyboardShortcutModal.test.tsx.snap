// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyboardShortcutModal renders correctly when isOpen is true 1`] = `
<div
  data-testid="keyboard_shortcut_modal"
>
  <button
    class="kbs__close__btn"
  >
    <svg
      fill="none"
      height="16"
      style="height: 23px; width: 23px;"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12.1835 11.4905C12.1835 11.6885 12.0845 11.8865 11.8794 12.0915C11.5824 12.3885 10.9814 12.3956 10.6774 12.0915L3.98812 5.2043C3.69114 4.90732 3.68407 4.30628 3.98812 4.00222C4.29218 3.69816 4.88615 3.69816 5.1902 4.00222L11.8794 10.8894C12.0774 11.0874 12.1764 11.2854 12.1835 11.4905Z"
        fill="#BCC2C4"
      />
      <path
        d="M12.1834 4.50427C12.1834 4.70226 12.0844 4.90025 11.9854 5.09824L5.29616 12.0845C4.90018 12.3814 4.39813 12.3885 4.09408 12.0845C3.79002 11.7804 3.79002 11.1864 4.09408 10.8824L10.7833 3.99515C11.0803 3.69816 11.6813 3.69109 11.9854 3.99515C12.0844 4.09414 12.1834 4.29213 12.1834 4.49012L12.1834 4.50427Z"
        fill="#BCC2C4"
      />
    </svg>
  </button>
  <div
    class="kbs__container light"
  >
    <div
      class="kbs__col light"
    >
      <div
        class="kbs__title"
      >
        For
         
        <span
          class="kbs__platform__name"
        >
          Windows
        </span>
      </div>
      <div
        class="kbs__list light"
      >
        <div
          class="kbs__list__item light"
        >
          <span>
            Show/hide UI
          </span>
          <span
            class="kbs__cmd__key"
          >
            Ctrl \\
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Play
          </span>
          <span
            class="kbs__cmd__key"
          >
            Ctrl P
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Open Side Bar Panel
          </span>
          <span
            class="kbs__cmd__key"
          >
            Ctrl T
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Fit View
          </span>
          <span
            class="kbs__cmd__key"
          >
            Ctrl 1
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Zoom In
          </span>
          <span
            class="kbs__cmd__key"
          >
            Ctrl +
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Zoom Out
          </span>
          <span
            class="kbs__cmd__key"
          >
            Ctrl -
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Layout Node
          </span>
          <span
            class="kbs__cmd__key"
          >
            Ctrl L
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Select All
          </span>
          <span
            class="kbs__cmd__key"
          >
            Ctrl A
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Deselect
          </span>
          <span
            class="kbs__cmd__key"
          >
            Alt C
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Close Node
          </span>
          <span
            class="kbs__cmd__key"
          >
            Ctrl W
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            File Name
          </span>
          <span
            class="kbs__cmd__key"
          >
            Ctrl R
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Save
          </span>
          <span
            class="kbs__cmd__key"
          >
            Ctrl S
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Delete
          </span>
          <span
            class="kbs__cmd__key"
          >
            Ctrl D
          </span>
        </div>
      </div>
    </div>
    <div
      class="kbs__col light"
    >
      <div
        class="kbs__title"
      >
        For
         
        <span
          class="kbs__platform__name"
        >
          MacOs
        </span>
      </div>
      <div
        class="kbs__list light"
      >
        <div
          class="kbs__list__item light"
        >
          <span>
            Show/hide UI
          </span>
          <span
            class="kbs__cmd__key"
          >
            ⌘ z
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Play
          </span>
          <span
            class="kbs__cmd__key"
          >
            ⌘ P
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Open Side Bar Panel
          </span>
          <span
            class="kbs__cmd__key"
          >
            ⌘ T
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Fit View
          </span>
          <span
            class="kbs__cmd__key"
          >
            ⌘ 1
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Zoom In
          </span>
          <span
            class="kbs__cmd__key"
          >
            ⌘ +
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Zoom Out
          </span>
          <span
            class="kbs__cmd__key"
          >
            ⌘ -
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Layout Node
          </span>
          <span
            class="kbs__cmd__key"
          >
            ⌘ L
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Select All
          </span>
          <span
            class="kbs__cmd__key"
          >
            ⌘ A
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Deselect
          </span>
          <span
            class="kbs__cmd__key"
          >
            Option C
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Close Node
          </span>
          <span
            class="kbs__cmd__key"
          >
            ⌘ W
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            File Name
          </span>
          <span
            class="kbs__cmd__key"
          >
            ⌘ R
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Save
          </span>
          <span
            class="kbs__cmd__key"
          >
            ⌘ S
          </span>
        </div>
        <div
          class="kbs__list__item light"
        >
          <span>
            Delete
          </span>
          <span
            class="kbs__cmd__key"
          >
            ⌘ D
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
